utilities_common/cli.py

```diff
@@ -662,3 +663,34 @@ def query_yes_no(question, default="yes"):
         else:
             sys.stdout.write("Please respond with 'yes' or 'no' "
                              "(or 'y' or 'n').\n")
+
+
+class UserCache:
+    """ General purpose cache directory created per user """
+
+    CACHE_DIR = "/tmp/cache/"
+
+    def __init__(self, app_name=None, tag=None):
+        """ Initialize UserCache and create a cache directory if it does not exist.
+
+        Args:
+            tag (str): Tag the user cache. Different tags correspond to different cache directories even for the same user.
+        """
+        self.uid = os.getuid()
+        self.app_name = os.path.basename(sys.argv[0]) if app_name is None else app_name
+        self.cache_directory_suffix = str(self.uid) if tag is None else f"{self.uid}-{tag}"
+        self.cache_directory_app = os.path.join(self.CACHE_DIR, self.app_name)
+        self.cache_directory = os.path.join(self.cache_directory_app, self.cache_directory_suffix)
+        os.makedirs(self.cache_directory, exist_ok=True)
```

Comment:
There would be an issue if the directory `/tmp/cache` or any subdirectory, such as `/tmp/cache/portstat` is created by root. 
For example, a user login as root, and run command `portstat`, then the owner of `/tmp/cache`  would be `root`
```
root@sonic:~# ls -al /tmp/|grep -w cache
drwxr-xr-x 3 root  root  4096 Jul 26 02:14 cache
root@sonic:~# ls -al /tmp/cache
total 12
drwxr-xr-x 3 root root 4096 Jul 26 02:14 .
drwxrwxrwt 1 root root 4096 Jul 26 02:14 ..
drwxr-xr-x 3 root root 4096 Jul 26 02:14 portstat
```
Then any non-root user will have no permission to show the counters
```
admin@sonic:~$ portstat 
Traceback (most recent call last):
  File "/usr/local/bin/portstat", line 640, in <module>
    main()
  File "/usr/local/bin/portstat", line 579, in main
    cache = UserCache(tag=tag_name)
  File "/usr/local/lib/python3.9/dist-packages/utilities_common/cli.py", line 684, in __init__
    os.makedirs(self.cache_directory, exist_ok=True)
  File "/usr/lib/python3.9/os.py", line 225, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 13] Permission denied: '/tmp/cache/portstat/1000'
```

Reply:
Thanks for catching this bug. To fix this I made cache and its subdirectories created with permissions to be writable for any user, so just like we have the whole ```/tmp``` writable for anyone, ```/tmp/cache``` is writable as well:

```
admin@r-panther-13:~$ ls -al / | grep tmp
drwxrwxrwt   1 root root 4096 Jul 26 10:29 tmp
admin@r-panther-13:~$ ls -al /tmp/ | grep cache
drwxrwxrwx 3 root  root   4096 Jul 26 09:34 cache
```